api/proto/teleport/userpreferences/v1/userpreferences.proto

```diff
@@ -51,6 +52,8 @@ message UpsertUserPreferencesRequest {
   UserPreferences preferences = 1;
   // username is the username of the owner of the user preferences to update.
   string username = 2;
+  // fieldmask is the field mask of the user preferences to update.
+  google.protobuf.FieldMask fieldmask = 3;
```

Comment:
Proto3 has `optional` functionality that allows you to generate `bool*` https://github.com/protocolbuffers/protobuf/blob/main/docs/field_presence.md#go-example

Reply:
Using optional will break the wire representation of the proto. 
It would probably require changing each field to optional  - our linter won't be happy about breaking the proto wire

Reply:
@ibeckermayer 

Adding new fields as optional doesn't break the wire representation but changing a field from non-optional to optional breaks.

My confusion started because @ryanclark wants to support field masks for every field - field mask was added to all fields in the web strct and if we gonna change everything to optional would break the wire representation.

If it's just adding a new field, I would recommend using a wrapper because optional creates the sense that the field is optional but it might be.

Reply:
I didn't want to introduce the quirk that some fields have to be optional whilst others aren't. We'd have to have optional bools and ints whilst other values aren't optional.

Putting a wrapper around a boolean again is requiring something special for that field, and also makes it exist in 3 states - undefined, true or false, which is something I wanted to avoid.

Additionally, an added benefit of using field masks means we can get rid of the manual merge code for merging user preferences.

Reply:
FWIW I think that field masks are the better solution for this problem, compared to optionals.

Reply:
@codingllama because why? 

As I see it, we're adding a lot of complexity including a recursive function that uses reflection to do something that can be trivially done by adding the `optional` keyword to some fields (perhaps requiring a breaking API change, but much simpler to understand and work with in the long run). 

I'm not taking the hard stance, granted I may be missing something, but under my current understanding it appears to be a mountain out of a molehill.

Reply:
`optional` isn't the answer here. We'd be making only `bool` and `int` fields in the user preferences response optional whereas the other fields wouldn't be. `optional` would make sense if this was only a message for updating the preferences, but it isn't.

This is the [recommended way](https://protobuf.dev/reference/java/api-docs/com/google/protobuf/FieldMask) of solving the issue we have, using a feature that is already in protobufs, that we [already are using in the codebase](https://github.com/gravitational/teleport/blob/64b3ccf793033190f39b2c8d3127fb1111276115/api/proto/teleport/devicetrust/v1/devicetrust_service.proto#L169).

> As I see it, we're adding a lot of complexity including a recursive function that uses reflection to do something

After speaking with @codingllama, the fieldmask implementation being in the backend is wrong, it should be the caller who specifies it, as it has the context. I had done this in the frontend originally and will be putting it back.

Reply:
@ibeckermayer I understand the concerns. The implementation can be much simpler, as Ryan also mentioned. If you look at the device trust usage, for example, you won't find reflective functions or external libraries, only a boring for-loop.

In terms of the design intention, which is also important here, field masks lets us cleanly deal with both primitives and nulls while keeping the regular proto "look". They are an established practice outside of Teleport and it's a worth tool to have in our toolbox. Granted, it's not always perfect, but it looks like a good fit here to me.

> it appears to be a mountain out of a molehill.

What a great expression, I'll steal this one for sure. :)

Reply:
> It's not exactly clear why this is a problem for the optional approach but not for the FieldMask. In either case the fact that some fields are optional remains the same.

They're not optional in the response when you get the user preferences. The default values are merged into any setting that is missing a value, so representing them as optional in protobuf is not true to the response type.

> This may be a recommendation elsewhere, but simply linking to the documentation on the technical specs for FieldMask doesn't demonstrate that. There's surely similar documentation on optional.

Sorry, I was meant to link to https://protobuf.dev/reference/protobuf/google.protobuf/#field-mask. "Recommended" is a bad word choice, I was meaning to relate the issue we have with the description of what field masks are for.

> That's certainly a valid point towards going in one particular direction, because all else equal it's simplest to have one cause to achieve the same effect. Though it looks like at this point we simply have that one case of using a FieldMask that you linked to, and [one other case](https://github.com/gravitational/teleport/blob/b5c11a29a7b9fbc427f3d9de921754da0b699b03/web/packages/teleterm/src/sharedProcess/api/proto/ptyHostService.proto#L77) that uses optional, so it's a wash on that front.

Could you expand on that example? I can't see `PtyEventExit` used in an update RPC and a get RPC but I'm not familiar with the code so I'm missing the similarity.

Reply:
> They're not optional in the response when you get the user preferences. The default values are merged into any setting that is missing a value, so representing them as optional in protobuf is not true to the response type.

Makes sense with your clarification.

> "Recommended" is a bad word choice, I was meaning to relate the issue we have with the description of what field masks are for.

Granted it's a common use case for `FieldMasks`.

> Could you expand on that example?

Not really, I just found the one example where we're using an `optional` field, I didn't check what it was being used for. So narrowed down to update ops it's an irrelevant point. It's clearly not being used for any sort of update operation, in fact afaict we're not using any of the `optional`ity of that field at all, maybe we can get rid of it 

(cc @gzdunek, looks like we only ever call `setSignal` which assumes a `number`, not `undefined`, though maybe there's some place where we use this `PtyEventExit` object's shape that I'm missing)

https://github.com/gravitational/teleport/blob/128878e1bb393e38cb38acc2ff20dc83c535b472/web/packages/teleterm/src/sharedProcess/ptyHost/ptyEventsStreamHandler.ts#L75

Reply:
We used `optional` field to align with `signal?: number` that is emitted from node-pty. Now I see that we used it incorrectly, `setSignal` should be called conditionally (probably I thought that I could pass `undefined` to `setSignal`).
Thanks for catching this @ibeckermayer.